h2o-core/src/main/java/water/H2O.java

```diff
@@ -1563,6 +1571,11 @@ static public void finalizeRegistration() {
   H2O( H2ONode[] h2os, int hash, int idx ) {
     _memary = h2os;             // Need to clone?
     java.util.Arrays.sort(_memary);       // ... sorted!
+    // create map key -> node
+    _memary_map = new ConcurrentHashMap<>();
+    for(H2ONode node: h2os){
+     _memary_map.put(node._key, node);
```

Comment:
This is suspicious - you are saying above that `_memary_map` contains " List of all clients that ever connected to this cloud" and here you are putting all nodes inside map. 

Furthermore, and this is more critical - you are duplicating state of `_memary` - that sounds like asking for problems in future...